### PR TITLE
Hive: Support dynamic partition insert

### LINKED ISSUE #3117: Hive: Support dynamic partition insert
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

sqlfluff currently handle only for static partition insert. So if sql uses dynamic partition insert, this makes error.

### Expected Behaviour

sqlfluff supports to parse dynamic partition insert

### Observed Behaviour

hql with dynamic partition insert
```
INSERT OVERWRITE TABLE aaa.bbb PARTITION (stats_id, log_date)
SELECT
    'a' as stats_id,
    'b' as log_date
FROM ccc.ddd
```

sqlfluff returns
```
 ~ >> sqlfluff lint de-scripts/test.hql                                                                
== [de-scripts/test.hql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'INSERT
                       | OVERWRITE TABLE aaa.bbb PARTITION...'
All Finished 📜 🎉!
```

hql with static partition insert
```
INSERT OVERWRITE TABLE aaa.bbb PARTITION (stats_id = 'a', log_date = 'b')
SELECT
    'a' AS a,
    'b' AS b
FROM ccc.ddd
```

sqlfluff returns
```
 ~ >> sqlfluff lint de-scripts/test.hql                                                                
All Finished 📜 🎉!
```

### How to reproduce

Add some sequence into `PartitionSpecGrammer` in `dialect_hive.py `
https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/dialects/dialect_hive.py#L123-L135

### Dialect

hive

### Version

sqlfluff, version 0.12.0
python 3.7.10

### Configuration

[sqlfluff]
dialect = hive
sql_file_exts = .sql,.hql
exclude_rules = L025,L031
templater = raw

[sqlfluff:indentation]
indented_using_on = true
template_blocks_indent = false

[sqlfluff:templater:jinja:context]
a=a

[sqlfluff:templater:python: context]
a=a

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 132
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = unqualified
unquoted_identifiers_policy = all

[sqlfluff:rules:L010]
capitalisation_policy = upper

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower
ignore_words = None

[sqlfluff:rules:L016]
ignore_comment_lines = True
ignore_comment_clauses = True

[sqlfluff:rules:L022]
comma_style = trailing

[sqlfluff:rules:L030]
extended_capitalisation_policy = upper
ignore_words = None

[sqlfluff:rules:L031]
force_enable = False

[sqlfluff:rules:L040]
capitalisation_policy = upper
ignore_words = None

[sqlfluff:rules:L042]
forbid_subquery_in = join

[sqlfluff:rules:L051]
fully_qualify_join_types = both

[sqlfluff:rules:L054]
group_by_and_order_by_style = explicit

[sqlfluff:rules:L063]
extended_capitalisation_policy = upper
ignore_words = None

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barunpuri**
### Brief summary of the change made

fixes #3117 
Support dynamic partition insert in hive dialect

### Are there any other side effects of this change that we should be aware of?
None

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Review by tunetheweb — CHANGES_REQUESTED** [review]

File: test/fixtures/dialects/hive/insert_overwrite_table_dynamic_partition.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+INSERT OVERWRITE TABLE foo
+PARTITION (a = 'test_foo', b) IF NOT EXISTS
+SELECT a, 'test_bar' AS b FROM bar;
```
[the remark this answers is not in the document]
Comment (barunpuri) [reply test/fixtures/dialects/hive/insert_overwrite_table_dynamic_partition.sql:3]: You mean test case 3 and 4? 
Case 1 already exists on examples
It's not hard to add case 3 but, is it okay to add grammatically incorrect code(case 4)?

[the remark this answers is not in the document]
Comment (tunetheweb) [reply test/fixtures/dialects/hive/insert_overwrite_table_dynamic_partition.sql:3]: I meant 1 and 3. OK just need to add 3 if 1 already exists.

4 you are correct you shouldn't add.

**Comment by tunetheweb** [comment]
Could we merge insert_overwrite_table_dynamic_partition_1.sql and insert_overwrite_table_dynamic_partition_2.sql? It's fine to have multiple SQL statements in the same file (semicolon-separated) - in fact it's preferred to keep related test case together!

**Comment by barunpuri** [comment]
> Could we merge insert_overwrite_table_dynamic_partition_1.sql and insert_overwrite_table_dynamic_partition_2.sql? It's fine to have multiple SQL statements in the same file (semicolon-separated) - in fact it's preferred to keep related test case together!

Of course, I merged examples. Thank you for your kindly review

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**